Pool add getDepositQuote, getWithdrawQuote api's

0xscuba opened the pull request:
### Context

* Currently, the client has to do math to figure out parameters for pool deposit and withdraw. We should provide functions that calculates these values

### Changes

* new pool api `getDepositQuote(maxTokenAIn, maxTokenBIn, slippage?)`
* new pool api `getWithdrawQuote(poolTokenIn, slippage?)`

### Usage Example

```typescript
const orca = getOrca(connection);
const pool = orca.getPool(OrcaPoolConfig.ORCA_SOL);

const orcaAmount = new Decimal(10);
const quote = await pool.getQuote(pool.getTokenA(), orcaAmount);
const solAmount = quote.getMinOutputAmount();
const poolTokenAmount = await pool.getDepositQuote(orcaAmount, solAmount); // <-- NEW

console.log(`Deposit at most ${orcaAmount} ORCA and ${solAmount} SOL, for at least ${
  poolTokenAmount
} pool tokens`);

const depositPayload = await pool.deposit(owner, orcaAmount, solAmount, poolTokenAmount);
const depositTxId = await depositPayload.execute();

console.log("Confirmation:", depositTxId);
```

odcheung commented on src/public/pools/types.ts:
```diff
@@ -69,6 +69,19 @@ export type OrcaPool = {
+    maxTokenAIn: Decimal | OrcaU64,
+    maxTokenBIn: Decimal | OrcaU64,
+    slippage?: Decimal
+  ) => Promise<OrcaU64>;
```
i think we should return a type of 
```
{
maxTokenAIn: OrcaU64,
maxTokenBIn: OrcaU64,
minPoolTokenAmountOUt: OrcaU64
}
```

This gives the users a chance to verify the adjusted maxTokenAIn / maxtokenBIn that were used to calculate the result of minPoolTokenAmount.

For example: 
User inputs maxUsdt: 4000, maxEth: 1 @ ETH price: 3500, slippage 1%
even though the user provided params still works, we can provide a more realistic maxUsdt value with this quote @ 3535
Guessing you can just use getQuote internally to do this.

0xscuba replied:
I think the information is definitely useful for the user but not sure if we should return the information from getDepositQuote for a few reasons:

1. We don't need to calculate tokenA/tokenB quote in order to calculate poolTokenAmount, because the calculation is independent of tokenA/tokenB price or amount. so it would be adding a calculation that isn't needed for the output
2. There was a user feedback that said getQuote method is slow
3. User can call getQuote themselves if they want to get a more tight bound on maxTokenAIn and maxTokenBIn value

0xscuba replied:
Actually iff we allow percentage as input for maxTokenAIn or maxTokenBIn, we should return maxTokenAIn, maxtokenBIn. But getQuote will not be used to calculate these values, but rather just percentage * amountUserOwns

odcheung commented on src/public/pools/types.ts:
```diff
@@ -90,6 +103,17 @@ export type OrcaPool = {
+   * @param poolTokenIn The amount of pool tokens to send in
+   * @param slippage The slippage in percentage you are willing to take in withdraw
+   */
+  getWithdrawQuote: (
+    poolTokenIn: Decimal | OrcaU64,
+    slippage?: Decimal
+  ) => Promise<{ minTokenAOut: OrcaU64; minTokenBOut: OrcaU64 }>;
+
```
People withdraw using either:
1. % of their deposit they want to pull out
2. have a specific amount of tokenX they want to pull out

To provide the same experience as our UI (on the UI people just drag the bar to get the # that they want), I think we should also be supporting the 2nd use-case in the sdk. 

Suggest we take in:

```
{
    constraintTokenAmt: Decimal | OrcaU64,
    constraintTokenMint: PubKey,
    slippage?: Decimal
}
```

and output the following that satisfies the user's constraint on the token min (tokenA, B, poolToken)
```
{
    minTokenAOut: OrcaU64,
    minTokenBOut: OrcaU64,
    maxPoolTokenAmountIn: OrcaU64
}
```

return undefined on constriantTokenAmt not being a mint of the pool | number not valid.

0xscuba replied:
* The % option would be really helpful so the user doesn't have to make the call to get amount of tokens they own manually! I'll also add this to deposit as well.
* Good call, I'll add two ways for getWithdrawQuote: (1) by pool token amount, (2) by one of the tokens

odcheung commented on src/public/pools/types.ts:
```diff
@@ -45,7 +45,7 @@ export type OrcaPool = {
     inputToken: OrcaToken,
     inputAmount: Decimal | OrcaU64,
-    slippage: Decimal
+    slippage?: Decimal
```
Add a comment giving the default slippage amount.

0xscuba commented:
@odcheung - Thanks for the review! Pushed new commits which makes these changes:

* Allow percentage based input for both getDepositQuote and getWithdrawQuote]
* Updated return values of both getDepositQuote and getWithdrawQuote
* Allow constraintToken based for getWithdrawQuote

odcheung commented on src/public/utils/web3/get-token-count.ts:
```diff
@@ -44,3 +47,73 @@ export async function getTokenCount(
+    return U64Utils.toPoolU64(tokenInAmount.value, pool, "tokenInAmount");
+  }
+
+  const totalUserTokenCount = await getUserTokenCount(
```
Let's keep the OrcaPool agnostic to the user wallet (lower the area of concern that orcapool owns). Any user-wallet type calculation should be done by the sdk user & the we should receive the raw token amount to process.

odcheung commented on src/public/pools/types.ts:
```diff
@@ -3,6 +3,47 @@ import Decimal from "decimal.js";
+/* WithdrawQuote types */
+export type WithdrawQuoteAmount = PoolTokenInAmount | ConstraintTokenInAmount;
+
+export type PoolTokenInAmount = {
+  poolTokenAmountIn: TokenInAmount;
+  type: "poolToken";
+};
```
I think we can keep the input type to constarintTokenType only and avoid type checking if we
- remove the % functionality
- expose the mint of the poolToken via OrcaPool.getPoolTokenMint()

If we do this, then all we need to do in our withdraw method is to 
1. match token by mint (A/B/LPToken)
2. derive LPAmountIn (what you have there is good). if doesn't match, throw error
3. withdraw

odcheung commented on src/public/pools/types.ts:
```diff
@@ -90,6 +145,18 @@ export type OrcaPool = {
   ) => Promise<TransactionPayload>;
 
+  /**
+   * Get minTokenAOut and mintTokenBOut amounts to be used for withdraw.
```
Update this line to your new input type:
Get suggested withdraw token amounts based on required withdraw amount of the pool token / one of the paired tokens

odcheung commented on src/public/pools/types.ts:
```diff
@@ -3,6 +3,47 @@ import Decimal from "decimal.js";
+  type: "poolToken";
+};
+
+export type ConstraintTokenInAmount = {
```
Having second thoughts about the naming of constraintToken... withdrawTokenAmount & withdrawTokenMint? 

In any case, remove "in" / "out" since it's different depending on which token you use.

0xscuba replied:
I like the `withdrawTokenAmount` & `withdrawTokenMint` :+1:

0xscuba commented:
Hey @odcheung, haha no worries! I really appreciate all the feedback, and sdk is better for it :)

Agree that we should be thoughtful around one-way-door decisions. I think in this case we definitely wanted to get it right because getDepositQuote, getWithdrawQuote will be an integral part of the SDK. Tangentially though, I think it's ok to push an api, then deprecate and add new versions of it without having api breaking changes (e.g. pool.someCoolMethod, later introduce pool.someCoolMethodV2).

Updated the PR with the following changes:

* Remove having different types of inputs
* Update comments, inputs, outputs
* New logic for calculating poolTokenIn_U64 using withdrawTokenAmount & withdrawTokenMint
* Also updated https://github.com/orca-so/typescript-sdk/pull/48 to match the changes

odcheung commented on src/public/pools/types.ts:
```diff
@@ -19,6 +31,12 @@ export type OrcaPool = {
   getTokenB: () => OrcaPoolToken;
 
+  /**
+   * Query the tokenMint public key of this pool.
```
nit: "query the mint public key for the pool token of this pool"

0xscuba commented:
@odcheung - thanks for the review! updated the commented, and merging now. I'll also bump the version and publish it
